src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java

```diff
@@ -152,39 +152,44 @@ public long add(ByteBuffer term, T key, InvalidVectorBehavior behavior)
             validateIndexable(vector, similarityFunction);
         }
 
-        var bytesUsed = new AtomicLong();
-        var newVector = new AtomicBoolean();
+        var bytesUsed = 0L;
+        VectorPostings<T> postings = postingsMap.get(vector);
         // if the vector is already in the graph, all that happens is that the postings list is updated
         // otherwise, we add the vector in this order:
-        // 1. to the vectorValues
-        // 2. to the postingsMap
+        // 1. to the postingsMap
+        // 2. to the vectorValues
         // 3. to the graph
         // This way, concurrent searches of the graph won't see the vector until it's visible
         // in the other structures as well.
-        var postings = postingsMap.computeIfAbsent(vector, v -> {
-            var bytes = RamEstimation.concurrentHashMapRamUsed(1); // the new posting Map entry
-            var ordinal = nextOrdinal.getAndIncrement();
```

Comment:
Why wasn't `computeIfAbsent` in conjunction with and `AtomicInteger` correct? I am trying to see it, but I haven't seen it yet.

Reply:
Found it. The `computeIfAbsent` in `ConcurrentSkipListMap` is called optimistically, which resulted in skipping ordinal values, and incorrect state in the rest of the map.

```java
    /**
     * If the specified key is not already associated with a value,
     * attempts to compute its value using the given mapping function
     * and enters it into this map unless {@code null}.  The function
     * is <em>NOT</em> guaranteed to be applied once atomically only
     * if the value is not present.
     *
     * @param key key with which the specified value is to be associated
     * @param mappingFunction the function to compute a value
     * @return the current (existing or computed) value associated with
     *         the specified key, or null if the computed value is null
     * @throws NullPointerException if the specified key is null
     *         or the mappingFunction is null
     * @since 1.8
     */
    public V computeIfAbsent(K key,
                             Function<? super K, ? extends V> mappingFunction) {
        if (key == null || mappingFunction == null)
            throw new NullPointerException();
        V v, p, r;
        if ((v = doGet(key)) == null &&
            (r = mappingFunction.apply(key)) != null)
            v = (p = doPut(key, r, true)) == null ? r : p;
        return v;
    }
```

Reply:
It's particularly interesting because a concurrent hashmap does apply the function atomically once.

Reply:
Ah, no wonder that tripped me up.  99% of the time I'm using CHM and I could have sworn this method made an exactly-once guarantee.  I feel slightly less dumb now.

Reply:
Me too! I was reading your solution and couldn't see how it fixed the concurrency bug.

Reply:
As a side benefit, I think the new code is less fragile without the special casing around `newVector`

Reply:
Definitely. One last nit is that since we aren't able to write a test for it, would you add a comment to prevent incorrect refactoring?

Reply:
commented and merged